src/Disks/ObjectStorages/DiskObjectStorageVFS.cpp

```diff
@@ -0,0 +1,114 @@
+#include "DiskObjectStorageVFS.h"
+#include "DiskObjectStorageVFSTransaction.h"
+#include "IO/WriteBufferFromFile.h"
+#include "Interpreters/Context.h"
+#include "ObjectStorageVFSGCThread.h"
+
+namespace DB
+{
+DiskObjectStorageVFS::DiskObjectStorageVFS(
+    const String & name_,
+    const String & object_storage_root_path_,
+    const String & log_name,
+    MetadataStoragePtr metadata_storage_,
+    ObjectStoragePtr object_storage_,
+    const Poco::Util::AbstractConfiguration & config,
+    const String & config_prefix,
+    zkutil::ZooKeeperPtr zookeeper_,
+    bool allow_gc_)
+    : DiskObjectStorage( //NOLINT
+        name_,
+        object_storage_root_path_,
+        log_name,
+        std::move(metadata_storage_),
+        std::move(object_storage_),
+        config,
+        config_prefix)
+    , allow_gc(allow_gc_)
+    , gc_thread_sleep_ms(config.getUInt64(config_prefix + ".object_storage_vfs_gc_period", 10'000))
+    , traits(VFSTraits{name_})
+    , zookeeper(std::move(zookeeper_))
+{
+    zookeeper->createAncestors(traits.log_item);
```

Comment:
We can download here the last snapshot and reconsile the list of things there with state on filesystem .
And / or we can create here migration script to convert existing s3 to vfs (by creating the first snapshot)